src/main/scala/chisel3/util/Decoupled.scala

```diff
@@ -227,20 +231,33 @@ class Queue[T <: Data](val gen: T,
   val do_deq = WireDefault(io.deq.fire())
 
   when (do_enq) {
-    ram(enq_ptr.value) := io.enq.bits
+    if (useSyncReadMem) {
+      ram.write(enq_ptr.value, io.enq.bits)
+    }
+    else {
+      ram(enq_ptr.value) := io.enq.bits
+    }
     enq_ptr.inc()
   }
   when (do_deq) {
     deq_ptr.inc()
+    deq_ptr_wire := Mux(deq_ptr.value === (entries.U - 1.U), 0.U, deq_ptr.value + 1.U)
   }
+    
   when (do_enq =/= do_deq) {
     maybe_full := do_enq
   }
 
   io.deq.valid := !empty
   io.enq.ready := !full
-  io.deq.bits := ram(deq_ptr.value)
-
+  
+  if (useSyncReadMem) {
+    io.deq.bits := ram.read(deq_ptr_wire)
+  }
```

Comment:
If `SyncReadMem` is introduced here, what's the behavior when `deq_ptr_wire === enq_ptr.value && do_enq`?
If this won't happen, you can add an assert here to placate users.
If this is going to happen, I think `ReadUnderWrite` behavior should also be specified.

Reply:
Sure, since it might happen, as per @albert-magyar's suggestion, the `WriteFirst` strategy has been adopted.